src/Symfony/Component/DependencyInjection/Compiler/ResolveBindingsPass.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\LazyProxy\ProxyHelper;
+
+/**
+ * @author Guilhem Niot <guilhem.niot@gmail.com>
+ */
+class ResolveBindingsPass extends AbstractRecursivePass
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function processValue($value, $isRoot = false)
+    {
+        if (!$value instanceof Definition || empty($value->getBindings())) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        $parameterBag = $this->container->getParameterBag();
+
+        if ($class = $value->getClass()) {
+            $class = $parameterBag->resolveValue($class);
+        }
+
+        $bindings = $value->getBindings();
+
+        $calls = $value->getMethodCalls();
+        // constructor
+        $calls[] = array(null, $value->getArguments());
+
+        foreach ($calls as $i => $call) {
+            list($method, $arguments) = $call;
+            $method = $parameterBag->resolveValue($method);
+            $parameters = null;
+            $resolvedArguments = array();
+
+            if (null === $reflectionMethod = $this->getReflectionMethod($class, $method)) {
+                continue;
+            }
+
+            foreach ($reflectionMethod->getParameters() as $key => $parameter) {
+                if (isset($arguments[$key])) {
+                    continue;
+                }
+
+                $typeHint = ProxyHelper::getTypeHint($reflectionMethod, $parameter, true);
+
+                if (!isset($bindings[$typeHint])) {
+                    continue;
+                }
+
+                $arguments[$key] = $bindings[$typeHint];
+            }
+
+            if ($arguments !== $call[1]) {
+                ksort($arguments);
+                $calls[$i][1] = $arguments;
+            }
+        }
```

Comment:
> $bindings should be allowed to contain $named args, so that it can be used with scalar/array params also.

Yes I have to do that :)

>  $bindings could also be validated: what about throwing when a key doesn't match any existing class/interface? That would help people spot typos easily.

What if the class/interface doesn't exist/is not loaded yet? Is this an unsupported case?